src/Common/FileCache.cpp

```diff
@@ -404,9 +406,43 @@ LRUFileCache::FileSegmentCell * LRUFileCache::addCell(
             "Cache already exists for key: `{}`, offset: {}, size: {}.\nCurrent cache structure: {}",
             keyToStr(key), offset, size, dumpStructureUnlocked(key, cache_lock));
 
-    auto file_segment = std::make_shared<FileSegment>(offset, size, key, this, state);
-    FileSegmentCell cell(std::move(file_segment), this, cache_lock);
+    auto skip_or_download = [&]() -> FileSegmentPtr
+    {
+        if (state == FileSegment::State::EMPTY)
```

Comment:
Let's also add a check for `enable_cache_hits_threshold != 0`, e.g. in case `enable_cache_hits_threshold == 0` we will not keep anything in records map and stash queue. It will make sense to do that if cache settings could be changed dynamically, but now they can't, so let's check it.